### PR TITLE
DomacaNaloga1_AleksaCirkovic

### PR DESCRIPTION
**Pull request by aleksacirkovic99**
Domaca Naloga 1 - Naravni Zlepki - Aleksa Ćirković

### CODE REVIEW

**Comment by aleksacirkovic99** [comment]
@mrcinv I invite you to my pull request

File: DomNal/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Naravni zlepek
+
+Aleksa Ćirković
+
+Opis naloge:
+  - Naloga je, da se poveže niz točk na grafu na gladek način. Obstaja več točk, vsaka s svojo X in Y vrednostjo, in je potrebno narisati gladko krivuljo, ki gre skozi vse te točke.
```
Comment (mrcinv) [inline DomNal/README.md:6]: V resnici ne gre za gladko krivuljo, ampak za funkcijo. Interpolacija z gladkimi krivuljami je sorodna tema, ki je precej bolj komplicirana.

File: DomNal/demo.jl [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#Aleksa Ćirković
+
+#Naloga je, da se poveže niz točk na grafu na gladek način. Obstaja več točk, vsaka s svojo X in Y vrednostjo, in je potrebno narisati gladko krivuljo, ki gre skozi vse te točke.
+
+#Začne se s seznamom točk, ki jih je potrebno povezati. Vsaka točka ima X in Y vrednost.
+
+#Za povezovanje teh točk z gladko krivuljo uporablja se "Naravni interpolacijski kubični zlepek". To pomeni, da med vsakim parom točk narišete mini-krivulje, ki se lepo prilegajo skupaj in ustvarijo eno gladko krivuljo.
```
Comment (mrcinv) [inline DomNal/demo.jl:7]: Lahko ste bolj konkretni:
`da med vsakim parom točk narišete mini-krivulje, ki ...`-> `da na vsakem intervalu [xi, x_i+1] narišete mini-krivulje oblike y=p_i(x), kjer je p_i(x) kubični polinom. Polinome p_i izberemo tako, da je funkcija krivulja gladka.`

File: DomNal/demo.jl [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+#Na koncu se nariše ta krivulja na grafu. Za boljšo preglednost tudi so označene originalne točke in uporabljene različne barve za prikaz različnih delov krivulje.
+
+#Struktura Zlepek je osnova. Hrani vse potrebne informacije o interpolacijskih točkah (x in y vrednosti) ter koeficiente (a, b, c, d), ki so potrebni za izračun kubičnih zlepkov.
```
Comment (mrcinv) [inline DomNal/demo.jl:11]: Lahko ste bolj konkretni! Kako so koeficienti (a, b, c, d) povezani s krivuljo? Kakšna je formula za krivuljo med dvema točkama.

File: DomNal/demo.jl [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+#Funkcija interpoliraj izračuna koeficiente za kubični zlepek na podlagi danih točk. Uporablja matriko A in vektor v za rešitev sistema enačb, ki določajo koeficiente zlepka. Ti koeficienti so ključni za risanje gladke krivulje skozi vse dane točke.
```
Comment (mrcinv) [inline DomNal/demo.jl:13]: Ta stavek sodi v dokumentacijo funkcije znotraj kode. Tule raje opišite algoritem, ki ga uporabite, da poiščete koeficiente zlepka.

File: DomNal/src/Domaca01.jl [hunk 5]
```diff
@@ -0,0 +1,88 @@
+module Domaca01
+
+export Zlepek, interpoliraj, vrednost, plotDN
+
+using LinearAlgebra
+using Plots
+
+struct Zlepek
```
Comment (mrcinv) [inline DomNal/src/Domaca01.jl:8]: Dodajte dokumentacijo v obliki [dokumentacijskega niza](https://docs.julialang.org/en/v1/manual/documentation/). Glejte tudi kontrolni seznam v https://ucilnica.fri.uni-lj.si/pluginfile.php/201576/mod_label/intro/domace_seznam.pdf?time=1713875640360

File: DomNal/src/Domaca01.jl [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    x::Vector{Float64}
+    y::Vector{Float64}
+    a::Vector{Float64}
+    b::Vector{Float64}
+    c::Vector{Float64}
+    d::Vector{Float64}
+end
+
+
+function interpoliraj(x::Vector{Float64}, y::Vector{Float64})::Zlepek
```
Comment (mrcinv) [inline DomNal/src/Domaca01.jl:18]: Podobno kot prej, dodajte dokumentacijski niz.

File: DomNal/src/Domaca01.jl [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  n = length(x) - 1
+  h = diff(x)
+
+  
+  A = zeros(n + 1, n + 1)
+  A[1, 1], A[end, end] = 1, 1
+  v = zeros(n + 1)
+  for i = 2:n
+      A[i, i] = 2 * (h[i - 1] + h[i])
```
Comment (mrcinv) [inline DomNal/src/Domaca01.jl:27]: Od kod pridejo te enačbe. Navedite vir ali pa opišite izpeljavo v poročilu.

File: DomNal/src/Domaca01.jl [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    d::Vector{Float64}
+end
+
+
+function interpoliraj(x::Vector{Float64}, y::Vector{Float64})::Zlepek
+  n = length(x) - 1
+  h = diff(x)
+
+  
+  A = zeros(n + 1, n + 1)
```
Comment (mrcinv) [inline DomNal/src/Domaca01.jl:23]: Matrika A nima veliko neničelnih elementov, zato je bolj uporabiti razpršene matrike (SparseArrays.jl)

File: DomNal/src/Domaca01.jl [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+  Zlepek(x, y, a, b, c[1:end-1], d)
+end
+
+
+function vrednost(z::Zlepek, x::Float64)::Float64
+  for i = 1:length(z.x)-1
+      if x >= z.x[i] && x <= z.x[i+1]
+          dx = x - z.x[i]
+          return z.a[i] + z.b[i]*dx + z.c[i]*dx^2 + z.d[i]*dx^3
```
Comment (mrcinv) [inline DomNal/src/Domaca01.jl:53]: Za računanje polinoma je bolje uporabiti Hornerjevo obliko 
`a + dx(b + dx*(c + dx*d)))`

File: DomNal/src/Domaca01.jl [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  end
+
+  a = y[1:end-1]
+
+  Zlepek(x, y, a, b, c[1:end-1], d)
+end
+
+
+function vrednost(z::Zlepek, x::Float64)::Float64
+  for i = 1:length(z.x)-1
```
Comment (mrcinv) [inline DomNal/src/Domaca01.jl:50]: Interval na katerem leži x lahko v logaritemskem času poiščemo z bisekcijo. Glejte funkcijo `searchsortedfirst`.

File: DomNal/test/runtests.jl [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+
+@testset "Interpolation Accuracy" begin
+    x_points = [1.0, 2.0, 3.0, 4.0]
+    y_points = x_points .^ 2  
+    spline = interpoliraj(x_points, y_points)
+    
+    for (x, expected_y) in zip(x_points, y_points)
+        @test vrednost(spline, x) ≈ expected_y atol=1e-6
+    end
```
Comment (mrcinv) [inline DomNal/test/runtests.jl:15]: dobro bi bilo dodati še teste za odvode in druge odvode. Pa tudi test, ki preizkusi izračun znotraj interpolacijskega intervala ne le v interpolacijskih točkah.

File: DomNal/test/runtests.jl [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+@testset "Boundary Conditions" begin
+  
+  x_points = [0.0, 0.5, 1.0, 1.5, 2.0]
+  y_points = [0.0, 0.25, 1.0, 2.25, 4.0]  
+  spline = interpoliraj(x_points, y_points)
+
+  
+  plotDN(spline)  
+  println("Manually verify boundary conditions in the plot.")
```
Comment (mrcinv) [inline DomNal/test/runtests.jl:26]: To ne sodi v `runtests.jl`, ki je mišljena, da se poganja brez človeške interakcije. Vsaj zaenrat ne, dokle AI ni še toliko dobra :-)

**Review by mrcinv — COMMENTED** [review]
@aleksacirkovic99 Naloga je pravilno rešena. Koda je dobro strukturirana. Testi in dokumentacija pa sta pomankljivi. V poročilu pogrešam izpeljavo sistema, s katerim ste poiskali koeficiente a, b, c, d. Prav tako pogrešam več testnih primerov, ki bi testirali izračune koeficientov, kot tudi izračune verdnosti med interpolacijskimi točkami.

Predlagam tudi, da za naslednjo nalogo ustvarite vejo in "pull request" na svojem repozitoriju (kako se to naredi, je opisano v https://ucilnica.fri.uni-lj.si/pluginfile.php/201576/mod_label/intro/domace_seznam.pdf?time=1713875640360). Oceno sem vpisal na učilnici. Oceno lahko tudi popravite, če upoštevate moje pripombe.
